nntrainer/tensor/float_tensor.cpp

```diff
@@ -9,38 +9,164 @@
  * @bug		No known bugs except for NYI items
  */
 
+#include <iomanip>
+#include <iostream>
+
+#include <blas_interface.h>
 #include <float_tensor.h>
+#include <util_func.h>
 
 namespace nntrainer {
 
 FloatTensor::FloatTensor(std::string name_, Tformat fm) :
   TensorBase(name_, fm, Tdatatype::FP32) {}
 
-void FloatTensor::allocate() {}
+/// @todo support allocation by src_tensor
+void FloatTensor::allocate() {
+  if (empty() || data)
+    return;
+
+  /// allocate new memory for the tensor data
+  MemoryData *mem_data;
 
-void FloatTensor::deallocate() {}
+  mem_data = new MemoryData((void *)(new float[dim.getDataLen()]{}));
+  data = std::shared_ptr<MemoryData>(mem_data, [](auto *mem_data) {
+    delete[] mem_data->template getAddr<float>();
+    delete mem_data;
+  });
 
-void *FloatTensor::getData() const { return (void *)0; }
+  offset = 0;
+  initialize();
+}
+
+void FloatTensor::deallocate() {
+  data = nullptr;
+  offset = 0;
+}
 
-void *FloatTensor::getData(size_t idx) const { return (void *)0; }
+void *FloatTensor::getData() const {
+  if (!data)
+    return nullptr;
 
-void *FloatTensor::getAddress(unsigned int i) { return (void *)0; }
+  data->validate();
+  return data->getAddr<float>() + offset;
+}
+
+void *FloatTensor::getData(size_t idx) const {
+  if (!data)
+    return nullptr;
+
+  data->validate();
+  return data->getAddr<float>() + offset + idx;
+}
+
+void *FloatTensor::getAddress(unsigned int i) {
+  size_t index = getIndex(batch(), channel(), height(), width());
+  if (i > index) {
+    return nullptr;
+  }
+  return &((float *)getData())[i];
+}
 
 const void *FloatTensor::getAddress(unsigned int i) const {
-  return (const void *)0;
+  size_t index = getIndex(batch(), channel(), height(), width());
+  if (i > index) {
+    return nullptr;
+  }
+  return &((float *)getData())[i];
 }
 
-void FloatTensor::setValue(float value) {}
+void FloatTensor::setValue(float value) {
+  float *data = (float *)getData();
+  std::fill(data, data + size(), value);
+}
 
 void FloatTensor::setValue(unsigned int batch, unsigned int c, unsigned int h,
-                           unsigned int w, float value) {}
+                           unsigned int w, float value) {
+  ((float *)getData())[getIndex(batch, c, h, w)] = value;
+}
 
-void FloatTensor::setZero() {}
+void FloatTensor::setZero() {
+  if (contiguous) {
+    sscal(size(), 0, (float *)getData(), 1);
+  } else {
+    /// @todo implement apply_i
+    // apply_i<float>([](float val) -> float { return 0; });
+    setValue(0);
+  }
+}
 
-void FloatTensor::initialize() {}
+/// @todo support additional initializer
+void FloatTensor::initialize() {
+  if (empty() || !isAllocated())
+    return;
+
+  switch (initializer) {
+  case Initializer::ZEROS:
+    setZero();
+    break;
+  case Initializer::ONES:
+    setValue(1.0f);
+    break;
+  default:
+    break;
+  }
+
+  putData();
+}
 
-void FloatTensor::initialize(Initializer init) {}
+void FloatTensor::initialize(Initializer init) {
```

Comment:
That's a good point!
There are two `initialize(Initializer init)` functions because each initialize() is different.
In order words, `FloatTensor::initialize()` and `HalfTensor::initialize()` use same functions such as setZero(), setValue(), setRandNormal(), etc. However, the implementation of these functions differs from each other. Therefore, to call each initialize method correctly, they are separated into two. I hope this clarifies :)